### PR TITLE
Moved all file handling into `sources`

### PR DESCRIPTION
**Pull request by omh**
Now there's one place for all the url, file and directory handling and the parser is much simpler and the api for loading a schema or query is very small and simple:

- `sources.LoadSchema(file|dir|url)`
- `sources.LoadQueries(schema, []string{file, file})`

The parser can now focus on the details of extracting info from a schema/query instead.

### CODE REVIEW

File: sources/sources_test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	})
+
+	t.Run("errors if file not found", func(t *testing.T) {
+		run(t, "testdata/file_not_found.gql", true)
+	})
+
+	t.Run("from a directory", func(t *testing.T) {
+		run(t, "testdata/queries", false)
+	})
+}
```
Comment (omh) [inline sources/sources_test.go:85]: Missing a test to make sure it errors out if a directory is not found

File: sources/sources.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	if s.IncludePrelude {
+		sources = append(sources, validator.Prelude)
+	}
+	return sources, nil
+
+}
+
+type directoryLoader struct{}
+
+func (l directoryLoader) Load(source string) ([]*ast.Source, error) {
```
Comment (uesteibar) [inline sources/sources.go:80]: this shouldn't be required anymore, right? we'll only get files here

File: introspection/introspection.go [hunk 3]
```diff
@@ -63,6 +63,9 @@ func fetch(url string) ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to download schema: %w", err)
 	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("got unknown http status code: %s", resp.Status)
```
Comment (nestor-sk) [inline introspection/introspection.go:67]: unknown or invalid?

File: parser/parser_test.go [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
 )
 
-func TestParseSchemaFile(t *testing.T) {
+func TestParseSchema(t *testing.T) {
 	is := is.New(t)
 
-	schema, err := ParseSchemaFile("testdata/schemas/with_deprecations.gql")
+	schema, err := parser.ParseSchema(
+		source(t, "testdata/schemas/with_deprecations.gql"),
+		validator.Prelude,
```
Comment (uesteibar) [inline parser/parser_test.go:18]: what is this `Prelude` thing?

File: sources/testdata/file_list.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+testdata/query.gql
+testdata/queries/subfolder/one.gql
```
Comment (nestor-sk) [inline sources/testdata/file_list.txt:2]: Is this used now in any test?

File: sources/sources.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+	return sources, nil
+}
+
+func loaderForSource(source string, prelude bool) (Loader, error) {
+	if strings.HasPrefix(source, "http://") || strings.HasPrefix(source, "https://") {
+		return httpLoader{}, nil
+	}
+
+	isDir, err := isDirectory(source)
```
Comment (nestor-sk) [inline sources/sources.go:128]: I believe that now only take a list of files, if you want go deeper you need to **. Can we remove all related to passing a directory?

**Review by nestor-sk — COMMENTED** [review]
some early quick comments, but looking good!

File: sources/sources.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	sources, err := loader.Load(source)
+	if err != nil {
+		return nil, err
+	}
+
+	return parser.ParseSchema(sources...)
+}
+
+// LoadQueries will load one or more queries from either a file or a directory
```
[the remark this answers is not in the document]
Comment (omh) [reply sources/sources.go]: Good point, updated the doc string! 🙇🏻

File: sources/sources.go (hunk 2 above)
Comment (omh) [reply sources/sources.go]: Indeed! I've removed all the directory loading stuff.

File: introspection/introspection.go (hunk 3 above)
Comment (omh) [reply introspection/introspection.go]: Good point. `invalid` is indeed more correct here. I've updated the error message.

File: sources/testdata/file_list.txt (hunk 5 above)
Comment (omh) [reply sources/testdata/file_list.txt]: Nope, good spot. 💥 deleted.

File: sources/sources.go (hunk 6 above)
Comment (omh) [reply sources/sources.go]: Yup, Unai pointed out this too. I've removed all the directory parsing. 💥 gone.

File: introspection/introspection.go [hunk 8]
```diff
@@ -63,6 +63,9 @@ func fetch(url string) ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to download schema: %w", err)
 	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("got invalid http status code: %s", resp.Status)
```
Comment (uesteibar) [inline introspection/introspection.go:67]: a niptick but: not invalid, right? just not successful 😄

File: parser/parser_test.go (hunk 4 above)
Comment (uesteibar) [reply parser/parser_test.go:18]: ahhhhh nice! thanks 🙇

**Review by uesteibar — APPROVED** [review]

File: introspection/introspection.go (hunk 8 above)
Comment (omh) [reply introspection/introspection.go]: How about now? :D

```go
return nil, fmt.Errorf("introspection query failed, got http status code: %s", resp.Status)
```

Comment (uesteibar) [reply introspection/introspection.go]: 👌

**Pull request merged**